lib/middleware/query.js

```diff
@@ -23,6 +24,9 @@ var qs = require('qs');
  */
 
 module.exports = function query(options) {
+  if (options && options.name === 'parseExtendedQueryString') {
+    deprecate('default extended query parser is deprecated, set explicitly to override');
```

Comment:
The message just says "set explicitly to override" as the way to get rid of the message, but the implementation seems to actually say you have to set it only to extended? The wording of the message is what we want to do, but it doesn't seem like the implementation matches that. For example, a user should be able to set it to their own custom query parser and since we are not deprecating that, it shouldn't print a deprecation.

Reply:
@dougwilson - pardon me, but I am not able to follow.

we are printing deprecation only when the parser is `parseExtendedQueryString` right? and that is hard-coded in the default configuration?

Is `set explicitly to override` phrase what you are looking for?

or

retain the message, but allow application to override query parser which leads to suppression of the warning?

Reply:
Oh, sorry, I think I read the logic wrong; so we are not _removing_ the extended option, simply changing the _default_ from extended to simple. The deprecation should show when the user doesn't pick one; if they picked one explicitly (even if it is extended because that's what they want/need) then it shouldn't show the warning. It's similar to `express.urlencoded()` today.

Reply:
@dougwilson - we are not even doing that in `4.x` right? i.e., `we are not removing the extended option, simply changing the default from extended to simple` - are we? From the conversation in the PR, I infer it as we are planning to do that only for 5, and deprecate it in 4.x?

If that is correct, sending a deprecation warning on `default extended query parser` on all control flow paths makes sense, for the benefit of the user awareness?

Reply:
The idea is to deprecate what we are changing. This doesn't seem to match that and is deprecating too much in this impl. As in it deprecated all usage of the extended parser, default or specific, yes? But in 5 (and 4) we still allow extended to be used when specifically specifically by the user. See for example the way extended default is deprecated in express.urlencoded()

Reply:
@dougwilson - I implemented your suggestion, pls have a look! the approach is:
 - it is given that the extended query parser is used in default.
 - assume the user may or may not make explicit call.
 - use a flag `defaultExtendedQueryParser` to track user setting
 - warn deprecation only in the absence of user setting
 - move the whole logic to `lib/application.js` to capture the user settings